### PR TITLE
1110 codeupdate- inband and Concurrent update rebase

### PR DESCRIPTION
**Pull request by sagisin**
This PR contains the following commits:

Concurrent code update support.
Bootside mapping
Support for boot types attributes
All the defect fixes for inband and out of band updates in pldm.
surveillance timer. 

TESTED: Tested by doing inband updates and out of band updates.

### CODE REVIEW

File: libpldmresponder/oem_handler.hpp [hunk 1]
```diff
@@ -116,18 +116,29 @@ class Handler : public CmdHandler
...
     virtual void processSetEventReceiver() = 0;
 
-    /** @brief Interface to monitor the surveillance pings from remote terminus
-     *
-     * @param[in] tid - TID of the remote terminus
-     * @param[in] value - true or false, to indicate if the timer is
-     *                   running or not
-     * */
-    virtual void setSurvTimer(uint8_t tid, bool value) = 0;
-
```
Comment (manojkiraneda) [inline libpldmresponder/oem_handler.hpp:126]: why is this removed here and again added below ?

File: oem/ibm/libpldmresponder/file_io_by_type.cpp [hunk 2]
```diff
@@ -461,11 +461,6 @@ std::unique_ptr<FileHandler> getHandlerByType(uint16_t fileType,
         {
             return std::make_unique<LicenseHandler>(fileHandle, fileType);
         }
-        case PLDM_FILE_TYPE_LID_RUNNING:
-        {
-            return std::make_unique<LidHandler>(fileHandle, false,
-                                                PLDM_FILE_TYPE_LID_RUNNING);
-        }
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/file_io_by_type.cpp:468]: why is running type removed.

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 3]
```diff
@@ -270,7 +371,7 @@ void CodeUpdate::setVersions()
                                     propertiesChanged(newImageId,
                                                       "xyz.openbmc_project."
                                                       "Software.Activation"),
-                                    [this](sdbusplus::message_t& msg) {
+                                    [this](sdbusplus::message::message& msg) {
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:374]: dont revert master here

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 4]
```diff
@@ -646,16 +646,13 @@ void pldm::responder::oem_ibm_platform::Handler::buildOEMPDR(
     static constexpr auto objectPath = "/xyz/openbmc_project/inventory/system";
     const std::vector<std::string> slotInterface = {
         "xyz.openbmc_project.Inventory.Item.PCIeSlot"};
+    buildAllCodeUpdateEffecterPDR(this, PLDM_OEM_IBM_ENTITY_FIRMWARE_UPDATE,
+                                  ENTITY_INSTANCE_0,
+                                  PLDM_OEM_IBM_BOOT_SIDE_RENAME, repo);
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:651]: this was added twice

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 5]
```diff
@@ -1047,43 +1059,6 @@ void Handler::processSetEventReceiver()
...
-    {
-        info(
-            "Failed to stop surveillance timer while remote terminus status is ‘{HOST_TRANST_OFF}’ with Terminus ID ‘{TID}’ ",
-            "HOST_TRANST_OFF", hostTransitioningToOff, "TID", tid);
-        startStopTimer(value);
-        pldm::utils::reportError(
-            "xyz.openbmc_project.PLDM.Error.setSurvTimer.RecvSurveillancePingFail");
-    }
-}
-
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1086]: why was all this code deleted and added back ? below

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 6]
```diff
@@ -219,6 +243,65 @@ class Handler : public oem_platform::Handler
                 processSAIUpdate();
             }
         });
+        updateBIOSMatch = std::make_unique<sdbusplus::bus::match::match>(
+            pldm::utils::DBusHandler::getBus(),
+            propertiesChanged("/xyz/openbmc_project/bios_config/manager",
+                              "xyz.openbmc_project.BIOSConfig.Manager"),
+            [this, codeUpdate](sdbusplus::message::message& msg) {
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:250]: message_t

File: softoff/softoff.cpp [hunk 7]
```diff
@@ -285,13 +285,27 @@ int SoftPowerOff::hostSoftOff(sdeventplus::Event& event)
...
+    if (requestHostTransition !=
+        "xyz.openbmc_project.State.Host.Transition.Off")
+    {
+        effecterState = PLDM_SW_TERM_GRACEFUL_RESTART_REQUESTED;
+    }
+    else
+    {
+        effecterState = PLDM_SW_TERM_GRACEFUL_SHUTDOWN_REQUESTED;
+    }
+
```
Comment (manojkiraneda) [inline softoff/softoff.cpp:302]: can you explain whats the use case behind this change ?

**Review by manojkiraneda — CHANGES_REQUESTED** [review]

File: libpldmresponder/bios_config.cpp [hunk 8]
```diff
@@ -523,6 +523,28 @@ void BIOSConfig::updateBaseBIOSTableProperty()
         auto method = bus.new_method_call(service.c_str(), biosConfigPath,
                                           dbusProperties, "Set");
         std::variant<BaseBIOSTable> value = baseBIOSTableMaps;
+#ifdef OEM_IBM
```
Comment (Pavithrab7) [inline libpldmresponder/bios_config.cpp:526]: can we avoid this #OEM_IBM?

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 9]
```diff
@@ -83,6 +85,23 @@ int CodeUpdate::setNextBootSide(const std::string& nextSide)
...
+        const auto& priorityValue = std::get<uint8_t>(priorityPropValue);
+        if (priorityValue == 0)
+        {
+            // Requested next boot side is already set
+            return PLDM_SUCCESS;
+        }
+    }
+    catch (const std::exception& e)
+    {
+        // Alternate side may not be present due to a failed code update
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:101]: do we need a info trace here?

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 10]
```diff
@@ -560,114 +743,33 @@ int processCodeUpdateLid(const std::string& filePath)
...
     }
-    else
+    catch (const std::exception& e)
     {
-        error("Error occurred during fork, error number - {ERROR_NUM}}",
-              "ERROR_NUM", errno);
-        return PLDM_ERROR;
+        error(
+            "InbandCodeUpdate: Failed to Assemble code update image ERROR={ERR_EXCEP}",
+            "ERR_EXCEP", e.what());
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:765]: only "e"

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 11]
```diff
@@ -560,114 +743,33 @@ int processCodeUpdateLid(const std::string& filePath)
...
+        setCodeUpdateProgress(false);
+        auto sensorId = getFirmwareUpdateSensor();
+        sendStateSensorEvent(sensorId, PLDM_STATE_SENSOR_STATE, 0,
+                             uint8_t(CodeUpdateState::FAIL),
+                             uint8_t(CodeUpdateState::START));
+        return -1;
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:771]: can we return PLDM_ERROR? or -1 is what teh caller API expects

File: oem/ibm/libpldmresponder/inband_code_update.hpp [hunk 12]
```diff
@@ -13,12 +13,20 @@ namespace responder
 
 static constexpr uint8_t pSideNum = 1;
 static constexpr uint8_t tSideNum = 2;
+static constexpr uint8_t maxVersionRetry = 20;
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/inband_code_update.hpp:16]: can we add a comment on what does this mean and why?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 13]
```diff
@@ -78,7 +78,7 @@ int pldm::responder::oem_ibm_platform::Handler::
 
 int pldm::responder::oem_ibm_platform::Handler::
     oemSetStateEffecterStatesHandler(
-        uint16_t entityType, uint16_t entityInstance, uint16_t stateSetId,
+        uint16_t entityType, uint16_t /*entityInstance*/, uint16_t stateSetId,
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:81]: why is instance ID commented?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 14]
```diff
@@ -891,13 +901,14 @@ void pldm::responder::oem_ibm_platform::Handler::_processSystemReboot(
...
+                    info("InbandCodeUpdate: Setting the one time APR policy");
                     dBusIntf->setDbusProperty(dbusMapping, value);
                 }
                 catch (const std::exception& e)
                 {
                     error(
-                        "Failure in setting one-time restore policy, unable to set property PowerRestorePolicy, error - {ERROR}",
-                        "ERROR", e);
+                        "Setting one-time restore policy failed, unable to set property PowerRestorePolicy ERROR={ERR_EXCEP}",
+                        "ERR_EXCEP", e.what());
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:911]: use newer way of lg2

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 15]
```diff
@@ -394,6 +477,12 @@ class Handler : public oem_platform::Handler
      */
     void setSurvTimer(uint8_t tid, bool value);
 
+    /** @brief Method to reset or stop the surveillance timer
+     *
+     *   @param[in] value - true or false, to indicate if the timer
+     *                     should be reset or turned off*/
+    void startStopTimer(bool value);
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:484]: do we need this to be public API? earlier i see it was a private API

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 16]
```diff
@@ -63,12 +63,37 @@ class Handler : public oem_fileio::Handler
...
+using OptionValue = std::variant<int64_t, std::string>;
+using Option = std::vector<std::tuple<OptionString, OptionValue>>;
+using BIOSTableObj =
+    std::tuple<AttributeType, ReadonlyStatus, DisplayName, Description,
+               MenuPath, CurrentValue, DefaultValue, Option>;
+using BaseBIOSTable = std::map<AttributeName, BIOSTableObj>;
+
+using PendingObj = std::tuple<AttributeType, CurrentValue>;
+using PendingAttributes = std::map<AttributeName, PendingObj>;
+static constexpr auto PLDM_OEM_IBM_FRONT_PANEL_TRIGGER = 32837;
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:84]: why is this PANEL State set added in the this PR?

Comment (Pavithrab7) [reply oem/ibm/libpldmresponder/oem_ibm_handler.hpp:84]: I think we want to add this - PLDM_OEM_IBM_ENTITY_FIRMWARE_UPDATE ?

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 13 above)
Comment (Pavithrab7) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:81]: so then we can remove that parameter right?

File: oem/ibm/libpldmresponder/inband_code_update.cpp (hunk 11 above)
Comment (Pavithrab7) [reply oem/ibm/libpldmresponder/inband_code_update.cpp:772]: can we do this change then?

File: common/utils.cpp [hunk 17]
```diff
@@ -835,5 +835,78 @@ std::vector<pldm::pdr::EffecterID>
...
+            method.append(
+                biosConfigIntf, "PendingAttributes",
+                std::variant<PendingAttributesType>(pendingAttributes));
+            bus.call_noreply(method);
+        }
+        catch (const sdbusplus::exception::SdBusError& e)
+        {
+            info(
+                "Error setting the bios attribute ERROR = {ERR_EXCEP} ATTRIBUTE= {DBUS_ATTR} ATTRIBUTE VALUE={BIOS_ATTR}",
+                "ERR_EXCEP", e.what(), "DBUS_ATTR", dbusAttrName.c_str(),
```
Comment (Pavithrab7) [inline common/utils.cpp:904]: "e"

File: oem/ibm/libpldmresponder/inband_code_update.cpp (hunk 3 above)
Comment (Pavithrab7) [reply oem/ibm/libpldmresponder/inband_code_update.cpp:375]: can we change this?

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

File: host-bmc/host_pdr_handler.cpp [hunk 18]
```diff
@@ -167,6 +168,10 @@ HostPDRHandler::HostPDRHandler(
             {
                 isHostRunning = true;
                 isHostOff = false;
+                if (oemPlatformHandler != nullptr)
```
Comment (manojkiraneda) [inline host-bmc/host_pdr_handler.cpp:171]: if(!oemPlatformHandler)

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 13 above)
Comment (manojkiraneda) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:81]: yes, lets clean it up if we are not using.

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 4 above)
Comment (manojkiraneda) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:651]: ack

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 19]
```diff
@@ -63,6 +63,25 @@ class Handler : public oem_fileio::Handler
...
+using OptionValue = std::variant<int64_t, std::string>;
+using Option = std::vector<std::tuple<OptionString, OptionValue>>;
+using BIOSTableObj =
+    std::tuple<AttributeType, ReadonlyStatus, DisplayName, Description,
+               MenuPath, CurrentValue, DefaultValue, Option>;
+using BaseBIOSTable = std::map<AttributeName, BIOSTableObj>;
+
+using PendingObj = std::tuple<AttributeType, CurrentValue>;
+using PendingAttributes = std::map<AttributeName, PendingObj>;
+static constexpr auto PLDM_OEM_IBM_ENTITY_FIRMWARE_UPDATE = 24577;
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:84]: can we use this from libpldm directly ?

File: softoff/softoff.cpp (hunk 7 above)
Comment (manojkiraneda) [reply softoff/softoff.cpp:302]: got it, so this is a pending part in upstreaming soft off ?

File: oem/ibm/libpldmresponder/inband_code_update.cpp (hunk 3 above)
Comment (sagisin) [reply oem/ibm/libpldmresponder/inband_code_update.cpp]: done

**Review by manojkiraneda — APPROVED** [review]

**Review by Pavithrab7 — APPROVED** [review]
